File: tools/afl/src/scell_seal_decrypt.c
/*
 * Copyright (c) 2019 Cossack Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <themis/themis.h>

#include "readline.h"

int main(void)
{
	themis_status_t err = THEMIS_SUCCESS;

	/*
	 * Read test data.
	 */

	uint8_t *master_key_bytes = NULL;
	size_t master_key_size = 0;

	if (read_line_binary(stdin, &master_key_bytes, &master_key_size))
	{
		return 1;
	}

	uint8_t *user_context_bytes = NULL;
	size_t user_context_size = 0;

	if (read_line_binary(stdin, &user_context_bytes, &user_context_size))
	{
		return 1;
	}

	uint8_t *message_bytes = NULL;
	size_t message_size = 0;

	if (read_line_binary(stdin, &message_bytes, &message_size))
	{
		return 1;
	}

	/*
	 * Try decrypting it.
	 */

	uint8_t *decrypted_bytes = NULL;
	size_t decrypted_size = 0;

	err = themis_secure_cell_decrypt_seal(
		master_key_bytes, master_key_size,
		user_context_bytes, user_context_size,
		message_bytes, message_size,
		NULL, &decrypted_size);

	if (err != THEMIS_BUFFER_TOO_SMALL)
	{
		return 2;
	}

	decrypted_bytes = malloc(decrypted_size);
	if (!decrypted_bytes)
	{
		return 2;
	}

	err = themis_secure_cell_decrypt_seal(
		master_key_bytes, master_key_size,
		user_context_bytes, user_context_size,
		message_bytes, message_size,
		decrypted_bytes, &decrypted_size);

	if (err != THEMIS_SUCCESS)
	{
		return 2;
	}

	return 0;
}

File: tools/afl/generate/scell_seal_decrypt.go
/*
 * Copyright (c) 2019 Cossack Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"encoding/binary"
	"fmt"
	"github.com/cossacklabs/themis/gothemis/cell"
	"io"
	"os"
)

func writeByteString(w io.Writer, str []byte) {
	binary.Write(w, binary.BigEndian, uint32(len(str)))
	w.Write(str)
}

func main() {
	args := os.Args

	if (len(args) != 3) && (len(args) != 4) {
		fmt.Printf("usage:\n\t%s <key> <message> [context]\n", args[0])
		os.Exit(1)
	}

	key := []byte(args[1])
	message := []byte(args[2])

	var context []byte

	if len(args) == 4 {
		context = []byte(args[3])
	}

	sc := cell.New(key, cell.ModeSeal)

	encrypted, _, _ := sc.Protect(message, context)

	writeByteString(os.Stdout, key)
	writeByteString(os.Stdout, context)
	writeByteString(os.Stdout, encrypted)
}
